---
title: "This section focuses on GSEA. Copying https://www.jianshu.com/p/82787ddada38"
date: "`r strftime(Sys.Date(), '%B %d %Y')`"
output: html_document
---

```{r setup, include=FALSE}


#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install(version = "3.14")
#BiocManager::install("clusterProfiler")
#BiocManager::install("DOSE")
#BiocManager::install("AnnotationHub")
BiocManager::install("biomaRt")


library("tidyverse")
library("clusterProfiler")
library("DOSE")
library("AnnotationHub")
library("biomaRt")
library("enrichplot")

library("dplyr")
library("plyr")
library("tidyverse")
library("ggplot2")
library("ggpubr")


```

First you need to load the settings and the script with the loading functions (you will need to have the `here`  package installed in your computer):

```{r load_settings}
# Load project settings
##source(here::here("settings.R"))

# Load helper functions
#source(file.path(functions.dir, "output.R"))

# settings.R didn't work on my computer
root.dir <- "/Users/catchia/_P4SBIP/p4sbip/Omics_data_updated/"
data.dir <- "/Users/catchia/_P4SBIP/p4sbip/Omics_data_updated/data/"
data.gsea.dir <- "/Users/catchia/_P4SBIP/p4sbip/Omics_data_updated/data/GSEA"
functions.dir <- "/Users/catchia/_P4SBIP/p4sbip/Omics_data_updated/functions/"

```

Then you can load all datasets into different data.frames with the following code:

```{r load_data}
# if you'd see this error: lazy-load database '/Library/Frameworks/R.framework/Versions/4.1/Resources/library/readr/R/sysdata.rdb' is corrupt
# restart the R session manually or via .rs.restartR() in command
genes_upregulated.dir <- file.path(data.dir, "GSEA/20220313_Upregulated_Genes.csv")
genes_downregulated.dir <- file.path(data.dir, "GSEA/20220313_Downregulated_Genes.csv")
df.gene.upregulated <- read.csv(genes_upregulated.dir)
df.gene.downregulated <- read.csv(genes_downregulated.dir)

#df.counts.raw <- read_output_csv("raw_counts_transcriptomics", .output.dir = data.dir)
#df.transcriptomics.design <- read_output_csv("transcriptomics_design", .output.dir = data.dir)
#df.gene.id.matching <- read_output_csv("gene_id_matching", .output.dir = data.dir)
#df.proteomics <- read_output_csv("proteomics", .output.dir = data.dir)
#df.functional.categories <- read_output_csv("functional_categories", .output.dir = data.dir)
#df.go.terms <- read_output_csv("go_terms", .output.dir = data.dir)

```

EnrichGO
```{r }
# enrichGO https://bioconductor.org/packages/release/bioc/manuals/clusterProfiler/man/clusterProfiler.pdf
# organism code: syz from https://www.genome.jp/kegg/catalog/org_list.html
# synechocystis sp. PCC 6803 genome https://www.genome.jp/entry/gn:T02468
# KEGG Identifier https://www.kegg.jp/kegg/kegg3.html

# Retrieve the ensembl info
ensembl <- useMart("ENSEMBL_MART_ENSEMBL", dataset = biomaRtOrg)


hub <- AnnotationHub()
q <- query(hub, "Synechocystis")
id <- q$ah_id[length(q)]
db.synechocystis <- hub[[id]]

erichgo_all <- enrichGO(gene = df.gene.upregulated,
                        OrgDb = db.synechocystis,
                        keyType = "ncbi-geneid", 
                        ont = "ALL",
                        pvalueCutoff = 0.01,
                        pAdjustMethod ="BH", 
                        universe,
                        minGSSize = 10, 
                        maxGSSize = 500, 
                        qvalueCutoff = 0.2
                        )

enrichKEGG_all <- enrichKEGG(gene = df.gene.upregulated$gene_name,
                        organism = "syn",
                        keyType = "kegg",
                        pvalueCutoff = 0.05,
                        pAdjustMethod ="BH", 
                        minGSSize = 10, 
                        maxGSSize = 500, 
                        qvalueCutoff = 0.2,
                        use_internal_data = FALSE
                        )

dotplot(enrichKEGG_all,showCategory = 20)

```
